NIFI-1678

markap14 opened the pull request:
Push heartbeats to ZooKeeper instead of NCM

mcgilman commented:
This is a resubmission of #317 which was closed when we removed a temporary 1.x branch which it was referencing. Please see that PR for additional commentary.

markap14 commented:
I updated the PR with a new commit to address an issue that I ran into. When we create 'container' nodes in ZooKeeper whose paths are not sensitive, we will mark the node as wide-open, rather than applying ACLs. This allows us to not have to worry about if we have state management and heartbeats using different ACLs.

Also removed use of username/passwords from ZooKeeper configuration. This was done because ZooKeeper indicates that these should be used only for local testing and not for production. In production, security should be administered via Kerberos. Username and password were always sent over the wire unencrypted. Additionally, we are now using Apache Curator to handle Leader Election, and Curator does not appear to support using digest auth via username/password.

markap14 commented:
I updated the PR with a new commit to address an issue that I ran into. When we create 'container' nodes in ZooKeeper whose paths are not sensitive, we will mark the node as wide-open, rather than applying ACLs. This allows us to not have to worry about if we have state management and heartbeats using different ACLs.

Also removed use of username/passwords from ZooKeeper configuration. This was done because ZooKeeper indicates that these should be used only for local testing and not for production. In production, security should be administered via Kerberos. Username and password were always sent over the wire unencrypted. Additionally, we are now using Apache Curator to handle Leader Election, and Curator does not appear to support using digest auth via username/password.

mcgilman commented:
Running the updated PR I ran into an issue when testing node disconnection while running an embedded ZK. It appears that by shutting down a majority of the nodes, we can no longer write the heartbeats and consequently the last node that was shut down still appears as connected.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster-protocol/src/main/java/org/apache/nifi/cluster/protocol/jaxb/message/AdaptedNodeConnectionStatus.java:
```diff
@@ -0,0 +1,63 @@
+    private Long connectionRequestTime;
+
+    public AdaptedNodeConnectionStatus() {
+    }
```
Any reason for this constructor?

markap14 replied:
Nope. Removing.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java:
```diff
@@ -0,0 +1,376 @@
+        if (future != null) {
+            future.cancel(true);
+        }
+    }
```
Unless I am missing something I don't see FlowEngine being shut down. This may prevent NiFi from exiting gracefully since there will be one living thread.

olegz replied:
Also, in the event ```client.close()``` results in exception 'future' will not be canceled. Perhaps wrapping in try/catch?

markap14 replied:
re: FlowEngine - threads are daemon so they won't prevent shut down. However, it should be shutdown anyway. Will address.
client.close() also shouldn't ever throw an Exception, but it is probably worth canceling the future in a finally block.

markap14 replied:
Actually in the refactoring i did, these issues are both gone as I removed both the future and the FlowEngine :)

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java:
```diff
@@ -0,0 +1,376 @@
+                clusterCoordinator.reportEvent(null, Severity.ERROR,
+                    "Received heartbeat from " + heartbeat.getNodeIdentifier() + " but failed to process heartbeat due to " + e);
+                logger.error("Failed to process heartbeat from {} due to {}", heartbeat.getNodeIdentifier(), e.toString());
+                logger.error("", e);
```
Can't we just do ```logger.error(e)```?

markap14 replied:
Unfortunately, the SLF4J api doesn't provide an 'error(Throwable)' method - you must provide a message.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java:
```diff
@@ -0,0 +1,376 @@
+        // Disconnect any node that hasn't sent a heartbeat in a long time (8 times the heartbeat interval)
+        for (final NodeHeartbeat heartbeat : latestHeartbeats.values()) {
+            final long maxMillis = heartbeatIntervalMillis * 1000L * 8;
+            final long threshold = latestHeartbeatTime - maxMillis;
```
Do those need to be recomputed for each iteration?

markap14 replied:
Nope, good catch.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java:
```diff
@@ -0,0 +1,376 @@
+            }
+
+            latestHeartbeats.put(heartbeatMsg.getHeartbeat().getNodeIdentifier(), StandardNodeHeartbeat.fromHeartbeatMessage(heartbeatMsg, lastModifiedTime));
+            logger.debug("Received heartbeat from Node {}", nodeId);
```
Do you think the above code (lines 341-348) could be moved inside of try block (after line 332) so you can remove continue? It may just be me, but it would be much easier to follow.

markap14 replied:
I believe this is no longer a concern in the new version of the code.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java:
```diff
@@ -0,0 +1,376 @@
+        } catch (final NoNodeException e) {
+            // node did not exist. Just return.
+            logger.debug("Attempted to remove heartbeat for Node with ID {} but no ZNode existed at {}", nodeId, nodeHeartbeatPath);
+            return;
```
Just making sure that this is intentional. If log level is not DEBUG then it's as if it never happened? I mean nothing at all in the logs.

markap14 replied:
That is correct - we are attempting to delete a heartbeat that no longer exists... so we consider it a 'success' and are simply logging at a debug level that it was already gone.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/NodeHeartbeat.java:
```diff
@@ -0,0 +1,64 @@
+import org.apache.nifi.cluster.protocol.NodeIdentifier;
+
+public interface NodeHeartbeat {
+    /**
```
In general I wonder if we need this and some other interfaces? Is there a plan to have multiple implementations and if so, how are they going to be configured?

markap14 replied:
I think we may well have 'dummy' implementations in unit tests, so I like the separation of the interface.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/manager/impl/WebClusterManagerCoordinator.java:
```diff
@@ -0,0 +1,164 @@
+            }
+        } catch (final Exception e) {
+            logger.error("Failed to request node {} disconnect from cluster due to {}", nodeId, e);
+            logger.error("", e);
```
Come to think of it, why are we logging it twice?

markap14 replied:
The first line logs a toString() of the Exception, whereas the second line logs the stack trace. Personally, I find the log messages easier to read this way when someone copies & pastes it, as they often do not look to grab a stack trace but if the toString() of the Exception is part of the original message, it makes it more clear.

olegz commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/manager/impl/WebClusterManagerCoordinator.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    @Override
+    public Set<NodeIdentifier> getNodeIdentifiers(final NodeConnectionState state) {
```
Not a big deal, but personally I would probably name the method 'extract/build....' since 'get' conventionally implies something that exists.

mcgilman commented:
Reviewing...

mcgilman commented:
Functionally this is much better than before. Now when running clustered and we start losing nodes we are able to see the current cluster participation. However, once we have lost the quorum we are no longer able to update the Primary Node. This behavior is expected, however. Once the quorum is re-established the Primary is updated accordingly. As we move towards zero master clustering, when there is no quorum we will return appropriate response codes. 

Looks good. Thanks!
